## Exploring German eBay Car Sales Data

The aim of this project is to analize the included car listings on Kleinanzeigen, a classifieds section of the Germany's eBay website.



In [2]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv',encoding = 'Latin-1', sep = ',') 

__The data dictionary provided with data is as follows:__

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [3]:
#Print a concise summary of a dataFrame.
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
#return 1st 5 rows of dataframe
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [5]:
#explore array of existing columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

__Observation__

Column names in camelcase: dateCrawled, offerType,  vehicleType, yearOfRegistration, powerPS, monthOfRegistration, dateCreated, postalCode, lastSeen.

In [6]:
#renaming columns
autos.rename(columns={"yearOfRegistration": "registration_year", "monthOfRegistration": "registration_month", "notRepairedDamage": "notRepairedDamage", "dateCreated": "ad_created"},inplace = True, errors="raise",)


In [7]:
# convert the column names from camelcase to snakecase

def modified_col (col):
    col = col.strip()#removes white space
    col = col.replace(' ','_') #remove whitespace, add '_'
    col = col.replace(')','')#removes ','
    col = col.replace('(','')
    col = col.replace(',','')

    col = col.lower()
    return col

new_columns = []
for new_col in autos.columns:
    clean_col = modified_col(new_col)
    new_columns.append(clean_col)

autos.columns = new_columns #Assigned the modified column names back to the DataFrame.columns


In [8]:
#Exploring different state of dataframe (first n rows)
autos.head(20)

datecrawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13  2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14  2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15  2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16  2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17  2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18  2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19  2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   

    seller offertype    price   abtest vehicletype  registration_year  \
0   privat   Angebot   $5,000  control         bus               2004   
1   privat   Angebot   $8,500  control   limousine               1997   
2   privat   Angebot   $8,990     test   limousine               2009   
3   privat   Angebot   $4,350  control  kleinwagen               2007   
4   privat   Angebot   $1,350     test       kombi               2003   
5   privat   Angebot   $7,900     test         bus               2006   
6   privat   Angebot     $300     test   limousine               1995   
7   privat   Angebot   $1,990  control   limousine               1998   
8   privat   Angebot     $250     test         NaN               2000   
9   privat   Angebot     $590  control         bus               1997   
10  privat   Angebot     $999     test         NaN               2017   
11  privat   Angebot     $350  control         NaN               2000   
12  privat   Angebot   $5,299  control  kleinwagen               2010   
13  privat   Angebot   $1,350  control   limousine               1999   
14  privat   Angebot   $3,999     test  kleinwagen               2007   
15  privat   Angebot  $18,900     test       coupe               1982   
16  privat   Angebot     $350     test       kombi               1999   
17  privat   Angebot   $5,500     test       coupe               1990   
18  privat   Angebot     $300  control         bus               1995   
19  privat   Angebot   $4,150  control         suv               2004   

      gearbox  powerps     model   odometer  registration_month fueltype  \
0     manuell      158    andere  150,000km                   3      lpg   
1   automatik      286       7er  150,000km                   6   benzin   
2     manuell      102      golf   70,000km                   7   benzin   
3   automatik       71    fortwo   70,000km                   6   benzin   
4     manuell        0     focus  150,000km                   7   benzin   
5   automatik      150   voyager  150,000km                   4   diesel   
6     manuell       90      golf  150,000km                   8   benzin   
7     manuell       90      golf  150,000km                  12   diesel   
8     manuell        0     arosa  150,000km                  10      NaN   
9     manuell       90    megane  150,000km                   7   benzin   
10 

__some column data is not in english__

In [9]:
#exploring descriptive statistics for all columns
autos.describe(include = 'all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand notrepaireddamage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

### Dataframe observation

__Columns to be dropped due to lack of or similar value:__
*  seller, offertype, nrofpictures, notrepaireddamage

__Columns that need more investigation.__
* price, vehicletype

__Examples of numeric data stored as text that needs to be cleaned__
* odometer, price

In [10]:
# Converting price and odometer to numeric types,updating names.
autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "").astype(int)
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

#renaming columns
autos.rename({"odometer":"odometer_km",
              "abtest":"a_b_test",
                "datecrawled":"date_crawled",
                "vehicletype":"vehicle_type",
                "fueltype":"fuel_type",
                "postalcode":"postal_code",
                "lastseen":"last_seen" 
             }, axis = 1, inplace = True)

#removing 'seller', 'offertype','nrofpictures'
autos.drop(['seller', 'offertype','nrofpictures','notrepaireddamage'], axis = 1, inplace = True)

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price a_b_test vehicle_type  registration_year    gearbox  powerps   model  \
0   5000  control          bus               2004    manuell      158  andere   
1   8500  control    limousine               1997  automatik      286     7er   
2   8990     test    limousine               2009    manuell      102    golf   
3   4350  control   kleinwagen               2007  automatik       71  fortwo   
4   1350     test        kombi               2003    manuell        0   focus   

   odometer_km  registration_month fuel_type       brand           ad_created  \
0       150000                   3       lpg     peugeot  2016-03-26 00:00:00   
1       150000                   6    benzin         bmw  2016-04-04 00:00:00   
2        70000                   7    benzin  volkswagen  2016-03-26 00:00:00   
3        70000                   6    benzin       smart  2016-03-12 00:00:00   
4       150000                   7    benzin        ford  2016-04-01 00:00:00   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

### Exploring odometer and price column

In [11]:
autos.describe()

price  registration_year       powerps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month   postal_code  
count        50000.000000  50000.000000  
mean             5.723360  50813.627300  
std              3.711984  25779.747957  
min              0.000000   1067.000000  
25%              3.000000  30451.000000  
50%              6.000000  49577.000000  
75%              9.000000  71540.000000  
max             12.000000  99998.000000

#### Analyzing the odometer_km columns

In [12]:
#view min/max/median/mean etc
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
autos['odometer_km'].value_counts().sort_index().head()

5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64

In [14]:
#prints shape(series)
autos['odometer_km'].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000, 100000,  40000])

__Observation__

No outliers found in odomenter_km column, the odometer ranges from 5000 km to  150000 km.

#### Exploring price column

In [15]:
#exploring unique values of price col/series
autos['price'].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [16]:
#identifying shape of series/col
autos['price'].unique().shape

(2357,)

In [17]:
#Generate round of descriptive statistics.
autos['price'].describe().round()

count       50000.0
mean         9840.0
std        481104.0
min             0.0
25%          1100.0
50%          2950.0
75%          7200.0
max      99999999.0
Name: price, dtype: float64

In [18]:
#sort price object by labels (along an axis) and unique counts, first 20 index.
autos['price'].sort_index(ascending = True).value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [19]:
#exploring price object with price less than $5
autos[autos["price"].between(0,5)]


date_crawled                                               name  \
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
55     2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
71     2016-03-28 19:39:35  Suche_Opel_Astra_F__Corsa_oder_Kadett_E_mit_Re...   
80     2016-03-09 15:57:57  Nissan_Primera_Hatchback_1_6_16v_73_Kw___99Ps_...   
87     2016-03-29 23:37:22                      Bmw_520_e39_zum_ausschlachten   
...                    ...                                                ...   
49884  2016-03-11 13:55:30               Audi_a6_2.5l__Schnaeppchen_nur_heute   
49943  2016-03-16 20:46:08                                         Opel_astra   
49960  2016-03-25 22:51:55            Ford_KA_zu_verschenken_***Reserviert***   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   

       price a_b_test vehicle_type  registration_year    gearbox  powerps  \
27         0  control          NaN               2005        NaN        0   
55         1     test          NaN               2017  automatik      224   
71         0  control          NaN               1990    manuell        0   
80         0  control        coupe               1999    manuell       99   
87         0  control          NaN               2000        NaN        0   
...      ...      ...          ...                ...        ...      ...   
49884      0     test        kombi               1999    manuell      150   
49943      0  control          NaN               2016    manuell      101   
49960      0  control   kleinwagen               1999    manuell       60   
49974      0  control       cabrio               1983    manuell       70   
49984      0     test          NaN               2000        NaN        0   

          model  odometer_km  registration_month fuel_type           brand  \
27          NaN       150000                   0       NaN            ford   
55     e_klasse       125000                   7    benzin   mercedes_benz   
71          NaN         5000                   0    benzin            opel   
80      primera       150000                   3    benzin          nissan   
87          5er       150000                   0       NaN             bmw   
...         ...          ...                 ...       ...             ...   
49884        a6       150000                  11    diesel            audi   
49943     astra       150000                   8    benzin            opel   
49960        ka       150000                   6    benzin            ford   
49974      golf       150000                   2    benzin      volkswagen   
49984       NaN       150000                   0       NaN  sonstige_autos   

                ad_created  postal_code            last_seen  
27     2016-03-27 00:00:00        66701  2016-03-27 18:45:01  
55     2016-03-06 00:00:00        22111  2016-03-08 05:45:44  
71     2016-03-28 00:00:00         4552  2016-04-07 01:45:48  
80     2016-03-09 00:00:00        66903  2016-03-09 16:43:50  
87     2016-03-29 00:00:00        82256  2016-04-06 21:18:15  
...                    ...          ...                  ...  
49884  2016-03-11 00:00:00        27711  2016-03-12 03:17:08  
49943  2016-03-16 00:00:00        89134  2016-03-17 19:44:20  
49960  2016-03-25 00:00:00        34355  2016-03-25 22:51:55  
49974  2016-03-20 00:00:00         8209  2016-03-27 19:48:16  
49984  2016-03-31 00:00:00        12103  2016-04-02 19:44:53  

[1583 rows x 16 columns]

__price column observation__

The length of price column is 2357, there are 1583 cars with a price of 0 to 5. Not sure if this was the initial price to start a bid as some of the cars listed are luxury cars as well. will explore the data another way to get more information.


In [20]:
#comparing the Standard Deviation vs Max price
autos[autos["price"].between(481104,99999999)]

date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   

          price a_b_test vehicle_type  registration_year    gearbox  powerps  \
514      999999     test        kombi               2009    manuell      101   
2897   11111111     test    limousine               1973    manuell       48   
7814    1300000  control        coupe               1992        NaN        0   
11137  10000000  control        coupe               1960    manuell      368   
22947   1234566  control        kombi               1999  automatik      190   
24384  11111111     test          NaN               1995        NaN        0   
27371  12345678  control          NaN               2017        NaN       95   
37585    999990     test    limousine               1985    manuell      111   
39377  12345678  control          NaN               2018    manuell       95   
39705  99999999  control    limousine               1999  automatik      224   
42221  27322222  control    limousine               2014    manuell      163   
43049    999999     test          bus               1981    manuell       70   
47598  12345678  control    limousine               2001    manuell      101   
47634   3890000     test        coupe               2006        NaN      799   

             model  odometer_km  registration_month fuel_type           brand  \
514          focus       125000                   4    benzin            ford   
2897        escort        50000                   3    benzin            ford   
7814           NaN        50000                  12       NaN  sonstige_autos   
11137          NaN       100000                   1    benzin  sonstige_autos   
22947          NaN       150000                   2    diesel             bmw   
24384          NaN       150000                   0       NaN      volkswagen   
27371        punto       150000                   0       NaN            fiat   
37585        jetta       150000                  12    benzin      volkswagen   
39377          v40       150000                   6       NaN           volvo   
39705     s_klasse       150000                   9    benzin   mercedes_benz   
42221           c4        40000                   2    diesel         citroen   
43049  transporter       150000                   1    benzin      volkswagen   
47598       vectra       150000                   3    benzin            opel   
47634          NaN         5000                   7       NaN  sonstige_autos   

                ad_created  postal_code            last_seen  
514    2016-03-17 00:00:00        12205  2016-04-06 07:17:35  
2897   2016-03-12 00:00:00        94469  2016-03-12 22:45:27  
7814   2016-04-04 00:00:00        60598  2016-04-05 11:34:11  
11137  2016-03-29 00:00:00        73033  2016-04-06 21:18:11  
22947  2016-03-22 00:00:00        17454  2016-04-02 03:17:32  
243

__Second Observation on price column__

Comparing the Standard Deviation price max price, it's obvious several low end cars that are overvalued or have wrong prices listed, e.g Ford_Focus_Turnier listed at 999999 or a Ford_Punto listed for 12345678 and Schlachte_Golf_3 valued at 11111111	 .


The price column has lots of missing or inconsistent information, I think it is best to make note of these errors and only update prices if necessary.

### Exploring date columns

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [21]:
#exploring columns:
#'date_crawled','ad_created','last_seen','registration_year'
autos[['date_crawled','ad_created','last_seen','registration_year']][0:10].info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   date_crawled       10 non-null     object
 1   ad_created         10 non-null     object
 2   last_seen          10 non-null     object
 3   registration_year  10 non-null     int64 
dtypes: int64(1), object(3)
memory usage: 448.0+ bytes


In [22]:
# To include missing values in the distribution and to use percentages instead of counts,
#chain the Series.value_counts(normalize=True, dropna=False) method.
#To rank by date in ascending order (earliest to latest), chain the Series.sort_index() method.
#calculate the distribution of values in the date_crawled, ad_created, and last_seen columns (all string columns) as percentages.

print(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index())


2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64


__date_crawled observation__

It seems the initial date crawled is in March 2016 to april and the percentages decline sharply within the last 3 crawled days 2016-04-05 to 2016-04-07,perhaps in corelation there was a sharpe decline as well.

In [23]:
print(autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())


2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64


__ad_created column observation__

Far fewer ads were created in 2015 than 2016, June - December 2015, similarly to the crawled dates, 2016-04-05 to 2016-04-07,perhaps in corelation to ad creatd dates, there was a sharpe decline as well.

In [24]:
print(autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())


2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64


__last_seen column observation__

There's a high volume of cars last seen over the past three days, 2016-04-05 to 2016-04-07,definitely in corelation with crawled dates and add created, however relation is unknown. It is not possible for a high volume in sales contributed to this.

In [25]:
#statistical description of df
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

__registration_year  column observation__

It seems that after 1999 there's a steady increase in vehicle registration throughout the decade. However the min date(1000) and max date (9999) suggest that the dates are inaccurate.

In [26]:
#Remove the values outside those upper and lower bounds using Series.value_counts(normalize=True).
autos['registration_year'].value_counts(normalize=True, ascending = True)

1952    0.00002
9996    0.00002
1001    0.00002
1929    0.00002
1931    0.00002
         ...   
2003    0.05454
2004    0.05474
1999    0.06000
2005    0.06030
2000    0.06708
Name: registration_year, Length: 97, dtype: float64

__registration_year observation__

The earliest valid registration year is more realisticly determined from 1950 - 2016, a car can't be first registered after the listing was seen.
        

Let's explore the brand column further to identify popular brands and any factors that may support demand e.g price, odometer, reg date etc. 

In [27]:
#Exploring brand column
#Identifying top car brands available

autos['brand'].value_counts(normalize = True, ascending = False)


volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
citroen           0.01402
smart             0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

In [34]:
#exploring brand column, aggregating with reg year column.
value_brand = autos['brand'].value_counts(normalize = True, ascending = False)
unique_brand = value_brand[value_brand < 20].index #Series.index attribute access labels

#empty dict to store aggregated data
reg_mean_agg = {}

for cars in unique_brand:
    selected_brand = autos[autos['brand'] == cars]
    mean_reg_yr = selected_brand['registration_year'].mean()
    reg_mean_agg[cars] = int(mean_reg_yr)
    
reg_mean_agg

{'volkswagen': 2004,
 'opel': 2006,
 'bmw': 2004,
 'mercedes_benz': 2005,
 'audi': 2004,
 'ford': 2003,
 'renault': 2003,
 'peugeot': 2004,
 'fiat': 2003,
 'seat': 2008,
 'skoda': 2006,
 'mazda': 2003,
 'nissan': 2003,
 'citroen': 2003,
 'smart': 2005,
 'toyota': 2004,
 'sonstige_autos': 2030,
 'hyundai': 2007,
 'volvo': 2000,
 'mini': 2008,
 'mitsubishi': 2001,
 'honda': 2002,
 'kia': 2007,
 'alfa_romeo': 2002,
 'porsche': 2001,
 'suzuki': 2004,
 'chevrolet': 2002,
 'chrysler': 2001,
 'dacia': 2010,
 'daihatsu': 2002,
 'jeep': 2025,
 'subaru': 2028,
 'land_rover': 2003,
 'saab': 2087,
 'daewoo': 2003,
 'trabant': 1971,
 'jaguar': 2001,
 'rover': 2001,
 'lancia': 2001,
 'lada': 2000}

__Brand column aggregation__

Luxury brands obviously have higher asking prices and  make up a high volume of available brands, BMW, Mercedes, Audi, Volvo are more expensive. Lada, Lania, Opel and Renault are much cheaper. 

Let's explore this further by exploring additinal aggregation to get more insight. 


In [ ]:
#Using PD.SERIES using previous library
#calculate the mean mileage for top brands(aggregation)

value_brand = autos['brand'].value_counts(normalize = True, ascending = False)
unique_brand = value_brand[value_brand < 20].index #Series.index attribute access labels

#empty dict to store aggregated data
mean_agg_km_dict = {}

for cars in unique_brand:
    selected_brand = autos[autos['brand'] == cars]
    mean_km = selected_brand['odometer_km'].mean()
    mean_agg_km_dict[cars] = int(mean_km)
    
#Convert dictionary to series objects, using the series constructor.
km_series = pd.Series(mean_agg_km_dict)
print(km_series)

In [ ]:
#Using PD.SERIES using previous library
#calculate the mean price of car brands(aggregation) 

value_brand = autos['brand'].value_counts(normalize = True, ascending = False)
unique_brand = value_brand[value_brand < 20].index #Series.index attribute access labels


#empty dict to store aggregated data
mean_agg_price_dict = {}

for cars in unique_brand:
    selected_brand = autos[autos['brand'] == cars]
    mean_price = selected_brand['price'].mean()
    mean_agg_price_dict[cars] = int(mean_price)
    
mean_agg_price_dict

#Convert dictionary to series objects, using the series constructor.
price_series = pd.Series(mean_agg_price_dict)
print(price_series)

In [ ]:
#Create a dataframe from  series object(km_series from above) 
#using the dataframe constructor.

price_df = pd.DataFrame(km_series, columns=['avg_odometer_km'])
price_df

__Brand odometer aggregation__

Expensive cars have the highest average odometer readings - alpha romeo, bmw, mercedes and volvo,these brands are still able to maintain higher asking prices.

Some cheaper cars have high odometer readings but still considerably lower gap overall, eg nissian, toyota, fiat and ford.

The average registration date for vehicles are in the same ball park from early to mid 2000s, so the odometer, asking price is pretty consistent.